consensus/hotstuff/cruisectl/block_rate_controller.go

```diff
@@ -17,32 +21,45 @@ import (
 	"github.com/onflow/flow-go/state/protocol"
 )
 
-// measurement represents one measurement of block rate and error.
+// measurement represents a measurement of error associated with entering view v.
 // A measurement is taken each time the view changes for any reason.
-// Each measurement measures the instantaneous and exponentially weighted
-// moving average (EWMA) block rates, computes the target block rate,
-// and computes the error terms.
+// Each measurement computes the instantaneous error based on the projected
+// and target epoch switchover times, and updates error terms.
 type measurement struct {
 	view            uint64    // v       - the current view
 	time            time.Time // t[v]    - when we entered view v
-	blockRate       float64   // r[v]    - measured instantaneous block rate at view v
-	aveBlockRate    float64   // r_N[v]  - EWMA block rate over past views [v-N, v]
-	targetBlockRate float64   // r_SP[v] - computed target block rate at view v
-	proportionalErr float64   // e_N[v]  - proportional error at view v
-	integralErr     float64   // E_N[v]  - integral of error at view v
-	derivativeErr   float64   // ∆_N[v]  - derivative of error at view v
+	instErr         float64   // e[v]    - instantaneous error at view v (seconds)
+	proportionalErr float64   // e_N[v]  - proportional error at view v  (seconds)
+	integralErr     float64   // I_N[v]  - integral of error at view v   (seconds)
+	derivativeErr   float64   // ∆_N[v]  - derivative of error at view v (seconds)
+
+	// informational fields - not required for controller operation
+	viewDiff uint64        // number of views since the previous measurement
+	viewTime time.Duration // time since the last measurement
 }
 
 // epochInfo stores data about the current and next epoch. It is updated when we enter
 // the first view of a new epoch, or the EpochSetup phase of the current epoch.
 type epochInfo struct {
-	curEpochFinalView        uint64
-	curEpochTargetSwitchover time.Time
-	nextEpochFinalView       *uint64
-	epochFallbackTriggered   *atomic.Bool
+	curEpochFirstView     uint64
+	curEpochFinalView     uint64    // F[v] - the final view of the epoch
+	curEpochTargetEndTime time.Time // T[v] - the target end time of the current epoch
+	nextEpochFinalView    *uint64
+}
+
+// targetViewTime returns τ[v], the ideal, steady-state view time for the current epoch.
+func (epoch *epochInfo) targetViewTime() time.Duration {
+	return time.Duration(float64(epochLength) / float64(epoch.curEpochFinalView-epoch.curEpochFirstView+1))
```

Comment:
this is basically a constant, any reason keeping it in function?

Reply:
When the calculation of the derived data is so inexpensive, my preference is usually to only store the minimal required data so that there is no opportunity for storing inconsistent data. The surface area here is small enough here that it doesn't really make a difference either way, though.